code/server/server.cpp

```diff
@@ -0,0 +1,178 @@
+#include <iostream>
+#include <yarp/os/all.h>
+
+using namespace std;
+using namespace yarp::os;
+
+#define FIFO_SIZE 10
+
+volatile int intFIFOtmp[FIFO_SIZE];
+volatile int intFIFOnum;
+
+class FIFO{
+private:
+	int intHead;
+	int intQueLen;
+	int intFIFO[FIFO_SIZE];
+public:
+	FIFO();
+	int enqueue(int intEnqData);
+	int dequeue();
+	int head_value();
+	void delete_element(int intNatural);
+	void FIFO_copy();
+};
+
+FIFO::FIFO()	//constructor
+{
+	intHead = 0;
+	intQueLen = 0;
+	for(int i=0;i<FIFO_SIZE;i++){
+		intFIFO[i]=0;
+	}
+}
+
+int FIFO::enqueue(int intEnqData)	//enqueue the argument "intEnqData"
+{
+	if(intQueLen < FIFO_SIZE){
+		intFIFO[(intHead + intQueLen) % FIFO_SIZE] = intEnqData;
+		intQueLen++;
+		FIFO_copy();	//copy the content of FIFO to global variable for monitoring
+		return true;
+	}else{
+		return false;
+	}
+}
+
+int FIFO::dequeue()		//dequeue the first element of FIFO
+{
+	if(intQueLen>0){
+		intQueLen--;
+		if(intHead+1 < FIFO_SIZE){
+			intHead++;
+		}else{
+			intHead = 0;
+		}
+		FIFO_copy();	//copy the content of FIFO to global variable for monitoring
+		return true;
+	}else{
+		return false;
+	}
+}
+
+int FIFO::head_value()	//return the value of the first element of FIFO
+{
+	return (intFIFO[intHead]);
+}
+
+void FIFO::delete_element(int intNatural)	//delete the element of FIFO corespoinding to the argument "intNatural"
+{
+	int intLen = intQueLen;
+	for(int i=0; i<intLen; i++){
+		if(intNatural == intFIFO[intHead]){
+			dequeue();
+		}
+		else{
+			enqueue(intFIFO[intHead]);
+			dequeue();
+		}
+	}
+}
+
+void FIFO::FIFO_copy()
+{
+	if(intQueLen>0){
+		for(int i = 0; i < intQueLen; i++){
+			intFIFOtmp[i] = intFIFO[i];
+		}
+		intFIFOnum = intQueLen;
+	}
+}
+
+class Thread1 : public RateThread{
+public:
+	Thread1(int r):RateThread(r){}
+	virtual bool threadInit()
+	{
+		cout << "Starting thread" << endl;
+		return true;
+	}
+
+	virtual void afterStart(bool s)
+	{
+		if(s){
+			cout << "Thread started successfully" << endl;
+		}else{
+			cout << "Thread didn't start" << endl;
+		}
+	}
+
+	virtual void run()
+	{
+   		Semaphore Sema;
+   		Sema.wait();	//wait update procedure in main function
+
```

Comment:
Mmm if I well understood the `wait()` method, since the semaphore is instantiated every time just before the method call (initializing the internal counter to 1 by default), the first call of `wait()` will just cause the counter to become 0 without getting stuck there (it's the only thread calling `wait()` on that specific semaphore so there are no other threads to wait for), am I wrong?
Anyway, another suggestion @kawasetsu, just to simplify your life while trying to share the same semaphore in all the critical sections, you don't really need to use the semaphore in the `main()` function, you could just instantiate a semaphore inside the FIFO class and use it in all the methods of that class which modify/display the FIFO array (there are no other methods outside that class that deal with such array, so you would be done)